### PR TITLE
Cast keep nullable

### LINKED ISSUE #7119: Compatibility of CAST operator.
**Use case**
CAST in ClickHouse work in different way than in other DBMS.

1. Nullable is the part of data type. When you cast Nullable(UInt8) to String, you will get String.
In other databases nullability is kept with CAST.

2. You can specify "conversion instruction" instead of a type name inside CAST.
Example: `CAST(x AS UNSIGNED)` will cast Int8 to UInt8, Int16 to UInt16 and so on.

**Describe the solution you'd like**
1. Implement `cast_keep_nullable` setting.
2. Add a notion of named conversion instructions.

### PR DESCRIPTION
**Pull request by Potya**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):

- Non-significant

Detailed description / Documentation draft:
Now cast kept nullability. Add cast_keep_nullable setting

### CODE REVIEW

**Comment by akuzm** [comment]
Please add a functional test for this to `0_stateless` (see the description here: https://clickhouse.tech/docs/en/development/tests/#functional-tests)

File: src/Functions/FunctionsConversion.h [hunk 1]
```diff
@@ -2406,6 +2407,11 @@ class CastOverloadResolver : public IFunctionOverloadResolverImpl
                 " Instead there is a column with the following structure: " + column->dumpStructure(),
                 ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
 
+        Settings set;
```
Comment (4ertus2) [inline src/Functions/FunctionsConversion.h:2410]: It dows not work this way. You should get settings with `context.getSettingsRef()` that allowed in `create(const Context &)`

**Comment by alexey-milovidov** [comment]
FYI This pull request is not merged.

**Comment by 4ertus2** [comment]
Closed in favor of #11733

**Pull request merged**